worldedit-core/src/main/java/com/sk89q/worldedit/function/operation/ForwardExtentCopy.java

```diff
@@ -394,17 +394,23 @@ public Operation resume(RunContext run) throws WorldEditException {
                 if (maskFunc != null) {
                     copy = new RegionMaskTestFunction(sourceMask, copy, maskFunc);
                 } else {
-                    copy = new RegionMaskingFilter(source, sourceMask, copy);
+                    copy = new RegionMaskingFilter(sourceMask, copy);
                 }
             }
-            if (copyingBiomes && (source.isWorld() || region instanceof FlatRegion)) {
-                copy = CombinedRegionFunction.combine(copy, new BiomeCopy(source, finalDest));
-            }
             ExtentTraverser<ParallelQueueExtent> queueTraverser = new ExtentTraverser<>(finalDest).find(ParallelQueueExtent.class);
             Extent preloader = queueTraverser != null ? queueTraverser.get() : source;
             blockCopy = new RegionVisitor(region, copy, preloader);
         }
 
+        if (copyingBiomes && (source.isWorld() || region instanceof FlatRegion)) {
+            BiomeCopy biomeCopy = new BiomeCopy(source, finalDest);
+            RegionFunction biomeFunction = sourceMask == Masks.alwaysTrue() ?
+                    biomeCopy : new RegionMaskingFilter(sourceMask, biomeCopy);
+            RegionVisitor regionVisitor = new RegionVisitor(region, biomeFunction, source);
+            Operations.completeBlindly(regionVisitor);
+            this.affectedBiomeCols += regionVisitor.getAffected();
+        }
```

Comment:
Is it not possible to combine the biome and block copy into one? This implementation will require all chunks to be loaded twice and thus could make the copy take up to 2x as long for a large region